File: backend/app/schemas/ai.py
"""API schemas for AI financial intelligence."""

from __future__ import annotations

from pydantic import BaseModel, Field


class CategoryItemSchema(BaseModel):
    transaction_id: str
    description: str
    category: str
    confidence: float
    signals: list[str] = Field(default_factory=list)


class AnomalyItemSchema(BaseModel):
    transaction_id: str
    anomaly_type: str
    severity: str
    message: str
    score: float
    details: dict = Field(default_factory=dict)


class FraudSchema(BaseModel):
    risk_score: float
    risk_level: str
    flags: list[str] = Field(default_factory=list)
    anomaly_count: int = 0


class CorrectionSchema(BaseModel):
    transaction_id: str | None = None
    field: str
    original: str
    corrected: str
    confidence: float
    reason: str


class SuggestionSchema(BaseModel):
    id: str
    severity: str
    title: str
    message: str
    action: str | None = None
    transaction_id: str | None = None


class ConfidenceSchema(BaseModel):
    overall: float
    ocr: float | None = None
    layout: float | None = None
    financial: float | None = None
    semantic: float | None = None
    factors: list[str] = Field(default_factory=list)


class CategorySpendSchema(BaseModel):
    category: str
    total_debit: float
    total_credit: float
    count: int
    percent_of_debit: float


class AiInsightsResponse(BaseModel):
    statement_id: str
    cached: bool = False
    confidence: ConfidenceSchema
    fraud: FraudSchema
    category_spend: list[CategorySpendSchema] = Field(default_factory=list)
    spending_insight: dict = Field(default_factory=dict)
    anomaly_count: int = 0
    suggestion_count: int = 0
    top_category: str | None = None


class AiCategoriesResponse(BaseModel):
    statement_id: str
    categories: list[CategoryItemSchema]
    cached: bool = False


class AiAnomaliesResponse(BaseModel):
    statement_id: str
    anomalies: list[AnomalyItemSchema]
    fraud: FraudSchema
    cached: bool = False


class AiConfidenceResponse(BaseModel):
    statement_id: str
    confidence: ConfidenceSchema
    corrections: list[CorrectionSchema] = Field(default_factory=list)
    cached: bool = False


class AiSuggestionsRequest(BaseModel):
    statement_id: str | None = None
    query: str | None = None


class AiSuggestionsResponse(BaseModel):
    statement_id: str
    suggestions: list[SuggestionSchema]
    corrections: list[CorrectionSchema] = Field(default_factory=list)
    cached: bool = False


class SemanticSearchResponse(BaseModel):
    statement_id: str
    query: str
    results: list[dict]

File: frontend/src/types/ai.ts
export interface CategoryItem {
  transaction_id: string
  description: string
  category: string
  confidence: number
  signals: string[]
}

export interface AnomalyItem {
  transaction_id: string
  anomaly_type: string
  severity: 'low' | 'medium' | 'high'
  message: string
  score: number
  details?: Record<string, unknown>
}

export interface FraudAssessment {
  risk_score: number
  risk_level: string
  flags: string[]
  anomaly_count: number
}

export interface ConfidenceBreakdown {
  overall: number
  ocr?: number | null
  layout?: number | null
  financial?: number | null
  semantic?: number | null
  factors: string[]
}

export interface CategorySpend {
  category: string
  total_debit: number
  total_credit: number
  count: number
  percent_of_debit: number
}

export interface SmartSuggestion {
  id: string
  severity: string
  title: string
  message: string
  action?: string | null
  transaction_id?: string | null
}

export interface SmartCorrection {
  transaction_id?: string | null
  field: string
  original: string
  corrected: string
  confidence: number
  reason: string
}

export interface AiInsightsResponse {
  statement_id: string
  cached: boolean
  confidence: ConfidenceBreakdown
  fraud: FraudAssessment
  category_spend: CategorySpend[]
  spending_insight: Record<string, unknown>
  anomaly_count: number
  suggestion_count: number
  top_category: string | null
}

export interface AiCategoriesResponse {
  statement_id: string
  categories: CategoryItem[]
  cached: boolean
}

export interface AiAnomaliesResponse {
  statement_id: string
  anomalies: AnomalyItem[]
  fraud: FraudAssessment
  cached: boolean
}

export interface AiSuggestionsResponse {
  statement_id: string
  suggestions: SmartSuggestion[]
  corrections: SmartCorrection[]
  cached: boolean
}

export interface SemanticSearchResult {
  transaction_id: string
  score: number
  description: string
  category: string
}
